### PR TITLE
2538 feat/self registration

### PR DESCRIPTION
**Pull request by aviveiros11**
#### Any background context you want to provide?
PR to allow self registration of users first logging on to the site.  Note: This does not include assigning an organization, that will be handled by #2539.  When testing, the tester will be able to reach the home page with a red banner saying the user is unaffiliated with an organization.  Due to the reCAPTCHA domain req's - this can only be tested on `localhost:8000` 

#### What's this PR do?
Includes a self registration form with google reCAPTCHA and email verification link.  

#### How should this be manually tested?
Please contact me for the reCAPTCHA secret key to be included in the tester's `local_untracked`.  Reconfigure the development settings by running `DJANGO_SETTINGS_MODULE=config.settings.dev`. After the dev server is running: 

Scenario 1
- Follow the create account workflow all the way through.  (the verification email will be sent to the console)
- Logout and login with the same credentials used to create the account. 

Scenario 2
- Verify error handling on the create an account form is comprehensive. 
 - un-check reCAPTCHA and try to log in.
 - try entering password values that don't meet authentication validation.
 - try to create an account that already exists.

Scenario 3
- After creating an account, try to login with the credentials not via the confirmation email. (should not be possible).

#### What are the relevant tickets?
#2538

### CODE REVIEW

File: seed/landing/forms.py [hunk 1]
```diff
@@ -23,3 +25,19 @@ class LoginForm(forms.Form):
...
         required=True
     )
+
+
+class CustomCreateUserForm(UserCreationForm):
+    class Meta:
+        model = SEEDUser
+        fields = ['username']
+        widgets = {
+            'username': forms.fields.TextInput(attrs={'placeholder': 'Email Address'})
```
Comment (Ryoken) [inline seed/landing/forms.py:35]: Would `forms.fields.EmailInput` work better than a `TextInput` here?

File: seed/landing/templates/landing/account_activation_invalid.html [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    <div class="page login">
+        <div class="page_section">
+            {% include "landing/_marketing_bullets.html" %}
+            <!-- END .section_marketing -->
+            <div class="section_forms">
+                <div class="form_title">
+                    <h2>{% trans "Thank you" %}</h2>
+
+                    <div class="signup_form">
+                        <p>{% trans "This email link is invalid." %}<br></p>
```
Comment (Ryoken) [inline seed/landing/templates/landing/account_activation_invalid.html:13]: Does this need to be added to lokalise?

File: seed/landing/templates/landing/account_activation_sent.html [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    <div class="page login">
+        <div class="page_section">
+            {% include "landing/_marketing_bullets.html" %}
+            <!-- END .section_marketing -->
+            <div class="section_forms">
+                <div class="form_title">
+                    <h2>{% trans "Thank you" %}</h2>
+
+                    <div class="signup_form">
+                        <p>{% trans "We've sent a sign in link to your email to access your account." %}<br>{% trans "You should be receiving it shortly." %}</p>
```
Comment (Ryoken) [inline seed/landing/templates/landing/account_activation_sent.html:13]: Same as before, does this still need to be added to localise?

File: seed/landing/templates/landing/create_account.html [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+{% block content %}
+    <div class="page login">
+        <div class="page_section">
+            {% include "landing/_marketing_bullets.html" %}
+            <!-- END .section_marketing -->
+            <div class="section_forms">
+                <div class="form_title">
+                    <h2>{% trans "Create Your Account" %}</h2>
+                    <form id="AuthNewPassword" class="signup_form" action="" method="post">
+                    {% csrf_token %}
```
Comment (Ryoken) [inline seed/landing/templates/landing/create_account.html:23]: Super minor but shouldn't this and line 27 be indented once more?

File: seed/landing/templates/landing/home.html [hunk 5]
```diff
@@ -20,9 +20,11 @@ <h2>{% trans "Log in to SEED Platform™" %}</h2>
                         <table class="signup_table">
                             {{ login_form.as_table }}
                         </table>
-                        <input type="submit" class="btn btn-primary" value="{% trans "Log In" %}">
-                        <input type="reset" class="btn btn-default" value="{% trans "Cancel" %}">
+                        <input type="submit" class="btn btn-primary" value="{% trans "Log In" %}"><br>
+                    <div>
+                        <a class="create_account" href="{% url "landing:create_account" %}">{% trans "Create an Account" %}</a>
```
Comment (Ryoken) [inline seed/landing/templates/landing/home.html:25]: "Create an Account" is not in lokalise yet, but "Create my account" is.

File: seed/landing/templates/landing/login.html [hunk 6]
```diff
@@ -15,9 +15,11 @@ <h2>{% trans "Log in to SEED Platform™" %}</h2>
                     <table class="signup_table">
                         {{form.as_table}}
                     </table>
-                    <input type="submit" class="btn btn-primary" value="{% trans "Log In" %}">
-                    <a href="{% url "landing:landing_page" %}" type="submit" class="btn btn-default">Cancel</a>
-                    <a class="forgot_password" href="{% url "landing:password_reset" %}">{% trans "Forgot password?" %}</a>
+                    <input type="submit" class="btn btn-primary" value="{% trans "Log In" %}"><br>
+                    <div>
+                        <a class="create_account" href="{% url "landing:create_account" %}">{% trans "Create an Account" %}</a>
```
Comment (Ryoken) [inline seed/landing/templates/landing/login.html:20]: Same as before, "Create an Account" is not in lokalise yet but "Create my account" is.

File: seed/tasks.py [hunk 7]
```diff
@@ -31,6 +31,45 @@
 logger = get_task_logger(__name__)
 
 
+def invite_new_user_to_seed(domain, email_address, token, user_pk, first_name):
+    """Send invitation email to newly created user from the landing page.
+    NOTE: this funciton is only used on the landing page because the user has not been assigned an organization
```
Comment (Ryoken) [inline seed/tasks.py:36]: Minor typo "funciton"

**Review by adrian-lara — APPROVED** [review]
Testing the different scenarios and cases within each looks good. Code changes look reasonable, but I'm not too familiar with Django templating. Probably worth another review from anyone more familiar with those.

**Comment by axelstudios** [comment]
Bug: if the user tries to create an account and uses any uppercase letters in their email address they won't be able to log in.  The backend needs to convert the email address to lowercase before saving to the `username` field in the database (it's fine for the `email` database field to have uppercase letters though).

We had to fix this in the past also with the create_default_user command: https://github.com/SEED-platform/seed/blob/6bb9407f5efdcff3913bbd77f52a0f5f39332a22/seed/management/commands/create_default_user.py#L55-L61

File: seed/landing/views.py [hunk 8]
```diff
@@ -157,8 +157,8 @@ def create_account(request):
             ''' End reCAPTCHA validation '''
             if result['success']:
                 user = form.save(commit=False)
+                user.username = user.username.lower()
                 user.is_active = False
-                user.save()
```
Comment (adrian-lara) [inline seed/landing/views.py]: Things don't seem to be working, and I suspect it's from this line. Should we still be saving/creating the user Object here?

**Review by adrian-lara — COMMENTED** [review]
I'm not getting confirmation emails (in my terminal). I think it may be from the line that I commented on specifically, but definitely worth you look over. 

Also, the position of the log in modal shifts down, off-center on page load: 
Screen Shot 2021-02-26 at 12 57 04 PM

File: seed/landing/views.py (hunk 8 above)
Comment (aviveiros11) [reply seed/landing/views.py]: 🙏  Thanks for catching this, I'm not sure how that got deleted...

File: seed/landing/views.py [hunk 9]
```diff
@@ -124,3 +133,69 @@ def signup(request, uidb64=None, token=None):
...
+                'response': recaptcha_response
+            }
+            data = urllib.parse.urlencode(values).encode()
+            req = urllib.request.Request(url, data=data)
+            response = urllib.request.urlopen(req)
+            result = json.loads(response.read().decode())
+            ''' End reCAPTCHA validation '''
+            if result['success']:
+                user = form.save(commit=False)
+                user.username = user.username.lower()
```
Comment (adrian-lara) [inline seed/landing/views.py:160]: I end up with a 500 when attempting to make a user with a "duplicate" username as an existing user (case-sensitive bug).
Screen Shot 2021-02-26 at 2 15 01 PM

Probably worth wrapping the `save` in a try-except and/or using the same pattern you use to get the Captcha and password errors through to the frontend.

File: seed/landing/views.py (hunk 8 above)
Comment (adrian-lara) [reply seed/landing/views.py]: I actually don't see the modal shift happening anymore 👍

**Review by adrian-lara — COMMENTED** [review]
Manual testing the different cases again - things look to be working as expected! One small potential change noted in the comment

**Review by adrian-lara — APPROVED** [review]
Changes look good to me! I no longer get a 500 when entering an "existing" user's email, and the scenarios all seem to still work as expected 💯

**Comment by aviveiros11** [comment]
@Ryoken @axelstudios do you guys think this is good to go?

**Review by axelstudios — APPROVED** [review]
Re-tested, looks good 👍

**Pull request merged**